### ISSUE TITLE
Different ring hash on masters and nodes

### ISSUE THREAD

**Issue by plmikhail** [leo-project/leofs#1076]
This happened after we take over the node

[System Confiuration]
-----------------------------------+----------
 Item                              | Value
-----------------------------------+----------
 Basic/Consistency level
-----------------------------------+----------
                    system version | 1.3.8
                        cluster Id | leofs_1
                             DC Id | dc_1
                    Total replicas | 3
          number of successes of R | 1
          number of successes of W | 1
          number of successes of D | 1
 number of rack-awareness replicas | 0
                         ring size | 2^128
-----------------------------------+----------
 Multi DC replication settings
-----------------------------------+----------
 [mdcr] max number of joinable DCs | 2
 [mdcr] total replicas per a DC    | 1
 [mdcr] number of successes of R   | 1
 [mdcr] number of successes of W   | 1
 [mdcr] number of successes of D   | 1
-----------------------------------+----------
 Manager RING hash
-----------------------------------+----------
                 current ring-hash | 265d7b4c
                previous ring-hash | 265d7b4c
-----------------------------------+----------

 [State of Node(s)]
-------+-------------------------+--------------+----------------+----------------+----------------------------
 type  |          node           |    state     |  current ring  |   prev ring    |          updated at
-------+-------------------------+--------------+----------------+----------------+----------------------------
  S    | node01@10.3.15.101      | running      | a01acd3c       | a01acd3c       | 2018-07-11 18:02:21 +0300
  S    | node05@10.3.15.105      | running      | a01acd3c       | a01acd3c       | 2018-07-12 16:51:07 +0300
  S    | node06@10.3.15.106      | running      | a01acd3c       | a01acd3c       | 2018-07-11 17:57:18 +0300
  S    | node07@10.3.15.107      | running      | a01acd3c       | a01acd3c       | 2018-07-11 17:57:57 +0300
  S    | node08@10.3.15.108      | running      | a01acd3c       | a01acd3c       | 2018-07-11 17:58:13 +0300
  S    | node09@10.3.15.109      | running      | a01acd3c       | a01acd3c       | 2018-07-11 21:16:18 +0300
  G    | gw01@10.3.15.11         | running      | a01acd3c       | a01acd3c       | 2018-07-11 18:06:06 +0300
  G    | gw02@10.3.15.12         | running      | a01acd3c       | a01acd3c       | 2018-07-11 19:57:51 +0300
  G    | gw05@10.3.15.15         | running      | a01acd3c       | a01acd3c       | 2018-07-14 15:01:22 +0300
-------+-------------------------+--------------+----------------+----------------+----------------------------

and now we have a lot of errors on new node node01@10.3.15.101
like that:
error.20180714.03.1:[E]	node01@10.3.15.101	2018-07-14 03:12:49.271345 -0400	1531552369	leo_storage_handler_object:put/2	388	[{from,gateway},{method,put},{key,<<"production-thumbnail/parts/used/53/12/41/80124171/1_116x88.jpg">>},{req_id,13646461},{cause,unavailable}]
and that:
error.20180714.02.1:[E]	node01@10.3.15.101	2018-07-14 02:25:53.886026 -0400	1531549553	leo_storage_handler_object:get/3	158	[{from,gateway},{method,get},{key,<<"production-watermark/parts/used/53/11/93/80119330/2.jpg">>},{req_id,96437690},{cause,unavailable}]

on other storage nodes we haven't "{cause,unavailable}" errors

and finally we decide restart leofs-storage service on node01, and now it not start with errors:
[E]	node01@10.3.15.101	2018-07-14 18:03:33.125645 +0300	1531580613	null:null	0	Error in process <0.5748.0> on node 'node01@10.3.15.101' with exit value:
{badarg,[{ets,insert,[leo_env_values,{{env,leo_redundant_manager,options},[{version,1},{cluster_id,leofs_1},{dc_id,dc_1},{n,3},{r,1},{w,1},{d,1},{bit_of_ring,128},{num_of_dc_replicas,1},{mdcr_r,1},{mdcr_w,1},{mdcr_d,1},{num_of_rack_replicas,0},{max_mdc_targets,2}]}],[]},{leo_misc,set_env,3,[{file,"src/leo_misc.erl"},{line,143}]},{leo_redundant_manager_api,set_options,1,[{file,"src/leo_redundant_manager_api.erl"},{line,203}]},{leo_storage_api,register_in_monitor_1,3,[{file,"src/leo_storage_api.erl"},{line,110}]},{leo_storage_api,register_in_monitor,2,[{file,"src/leo_storage_api.erl"},{line,74}]}]}
[E]	node01@10.3.15.101	2018-07-14 18:03:33.649923 +0300	1531580613	leo_storage_app:after_proc_1/3	235	[{cause,"Not initialize leo_backend_db_sup"}]
[E]	node01@10.3.15.101	2018-07-14 18:03:33.650194 +0300	1531580613	null:null	0	{module,"leo_storage_app"},{function,"after_proc_1/3"},{line,224},{body,"Could NOT start backend-db sup"}
[E]	node01@10.3.15.101	2018-07-14 18:03:33.650682 +0300	1531580613	leo_storage_app:start	0	CRASH REPORT Process <0.44.0> with 0 neighbours exited with reason: bad return value ok from leo_storage_app:start(normal, []) in application_master:init/4 line 134

**Comment by plmikhail** [comment]
[manager ring dump](https://www.dropbox.com/s/ul1r263dyvts7ki/ring_cur.dump.63698805473?dl=0)
[node05 ring dump](https://www.dropbox.com/s/f1xut1pzpz268q7/ring_cur.dump.63698805628?dl=0)

**Comment by mocchira** [comment]
@plmikhail Thanks for reporting this issue and sorry for inconvenience.

> This happened after we take over the node

Can I ask you what procedure you took for the take over in details?
The terminal log or a script you'd use for this operation would be helpful to reproduce this issue on our environment.
and also please share the previous cluster status before node01@10.3.15.101 joined in and before the old node (being replaced by node01) still had joined in if you have those information.

and just in case, let me ask you that currently your services depending on LeoFS has not been affected by this problem as other nodes seem to work fine in accordance with your consistency setting (N=3, R=W=D=1)?
Let us know if any service has been affected, we will do our best to solve your problem ASAP.

**Comment by mocchira** [comment]
@plmikhail Once you provide the information I asked on the prev comment and it turns out that manager's ring hash is the latest (proper) one, I will ask you to issue the commands below

```shell
## $NodeList contains all your storage nodes and gateway nodes
for N in $NodeList do
  leofs-adm recover-ring $N
done
```

this would probably make the ring hash of all nodes in the cluster sync with the latest one and solve your problem.

**Comment by plmikhail** [comment]
@mocchira Thanks for the answer

> Can I ask you what procedure you took for the take over in details?

Procedure from here: https://leo-project.net/leofs/docs/admin/system_operations/cluster/
- Execute `leofs-adm detach` command to remove a target node in the cluster
- Then launch a new node to take over the detached node
- Finally, execute `leofs-adm reebalance` command to start rebalancing data in the cluster

>  I will ask you to issue the commands below

Result:
```
node01 - [ERROR] Could not connect
node05 - OK
node06 - OK
node07 - OK
node08 - OK
node09 - OK
gw01 - OK
gw02 - OK
gw05 - OK
```
nothing changed
manager ring: 265d7b4c
nodes and gateways ring: a01acd3c

> and also please share the previous cluster status before node01@10.3.15.101 joined in and before the old node (being replaced by node01) still had joined in if you have those information.

Before last operation (then we replaced node03 by node01) everything works fine.
After rebalance operation was started, we got different ring hash and a lot of errors on node01 (actually 80% of put operations didn't working).
We wanted to wait until the end of balancing before doing something.
Then rebalance operation completed and nothing changed we suspended node01 and restart leo-storage service on it.
Now put operations works, but we have different ring hash and can't get back node01.

**Comment by vstax** [comment]
I'd like to add a few comments on the situation. First of all, about main problem (RING). Dumps are identical on all nodes and gateways, _cur* files are identical to _prv* as well. Full dumps are here: [node05-ring.tar.gz](https://github.com/leo-project/leofs/files/2198046/node05-ring.tar.gz)

(a minor nitpick: it would be nice if timestamp was fixed for all the dumps, right now it can be "members_cur.dump.63698805473" and "ring_cur_worker.log.63698805474" for example)
On both managers _prv* are identical to _prv* and _cur* on nodes, and only "ring_cur.dump" differs. "ring_cur_worker.log" is the same across all nodes and managers. Full dumps from manager are here: [manager-ring.tar.gz](https://github.com/leo-project/leofs/files/2198045/manager-ring.tar.gz)

Restarting all nodes and manager doesn't change anything.

Second, the whole situation is as follows. The cluster originally was created with storage servers in VMs which were gradually replaced by physical servers. Physical servers and VMs had the same configuration (6 AVS directories, 16 containers in each). Basically what was done is
```
detach node06@10.3.15.106
(node node06 launched)
rebalance
rebalance
detach node05@10.3.15.105
(node node05 launched)
rebalance
rebalance
detach node04@10.3.15.104
(same)
rebalance
rebalance
detach node03@10.3.15.103
(same)
rebalance
rebalance
detach node02@10.3.15.102
(same)
rebalance
rebalance
detach node01@10.3.15.101
(same)
rebalance
rebalance
detach node03@10.3.15.103
(same)
rebalance
detach node04@10.3.15.104
(same)
rebalance
detach node02@10.3.15.102
(same)
rebalance
detach node01@10.3.15.101
(same)
rebalance
detach node03@10.3.15.101
(same)
rebalance
```
Where physical nodes were launched without data and usually with the same names, but in certain cases with different ones and data migration was performed by rebalance. Names like "node07" and "node08" appeared in cluster like that (I don't have exact information on what replaced what and when, @plmikhail might have this information, if it's important). Takeover of each next node was performed only after rebalance completely finished and queues were empty.
At this point, node09 is the last node running in VM, it will be migrated as soon as current situation is resolved.

The last operation was as described above, detachment of node03, which was replaced by node01. Around that time most of  writes started to fail, no errors on gateway. On nodes, especially node01 there was massive amount of "cause,unavailable" errors from leo_storage_handler_object:get/1. However, with R=1 all GET operations succeeded. At this point problem with RING on managers was noticed - I tried to help by checking with "whereis" how new objects should go to new nodes and found that this operation to be non-functional because of wrong results, and "node07" and "node09" never being present in output of "whereis" command.

Third, about problem with node01 that's not starting. It appeared later, during attempts to fix the situation somehow. It was noticed that when node01 is down, the cluster is performing much better than with it's available. Even just executing "suspend node01" on manager helps and writes start to succeed.  However, sometime after that (after being suspended) it refused to start. At the very end of startup, after all containers are opened, these lines appear in info logs
```
[I]     node01@10.3.15.101      2018-07-14 17:44:01.73437 +0300 1531579441      leo_storage_handler_del_directory:terminate/2   232     [{"terminate with reason",{noproc,{gen_server,call,[undefined,{get,<<"develop/test/27.mp4\n4\n">>},30000]}}}]
[I]     node01@10.3.15.101      2018-07-14 17:44:01.106658 +0300        1531579441      leo_storage_handler_del_directory:terminate/2   232     [{"terminate with reason",{noproc,{gen_server,call,[undefined,{get,<<"develop/test/26.mp4\n3\n">>},30000]}}}]
```
And in error logs
```
[E]	node01@10.3.15.101	2018-07-14 17:44:01.74042 +0300	1531579441	gen_server:call	0	gen_server leo_storage_handler_del_directory terminated with reason: no such process or port in call to gen_server:call(undefined, {get,<<"develop/test/27.mp4\n4\n">>}, 30000) in gen_server:call/3 line 212
[E]	node01@10.3.15.101	2018-07-14 17:44:01.77938 +0300	1531579441	gen_server:call	0	CRASH REPORT Process leo_storage_handler_del_directory with 0 neighbours exited with reason: no such process or port in call to gen_server:call(undefined, {get,<<"develop/test/27.mp4\n4\n">>}, 30000) in gen_server:terminate/7 line 812
[E]	node01@10.3.15.101	2018-07-14 17:44:01.78518 +0300	1531579441	gen_server:call	0	Supervisor leo_storage_sup had child leo_storage_handler_del_directory started with leo_storage_handler_del_directory:start_link() at <0.5420.0> exit with reason no such process or port in call to gen_server:call(undefined, {get,<<"develop/test/27.mp4\n4\n">>}, 30000) in context child_terminated
```
The end result is as you can see in the logs in the first post, it dies with "Could NOT start backend-db sup".

I've tried clearing queues on node01 but it does not help it to start. There are "delete" messages in queues of other nodes, e.g.
```
[root@leofs-mgr01 manager-ring]# leofs-adm mq-stats node09@10.3.15.109
              id                |    state    | number of msgs | batch of msgs  |    interval    |                 description                 
--------------------------------+-------------+----------------|----------------|----------------|---------------------------------------------
 leo_async_deletion_queue       |   running   | 211            | 1600           | 500            | async deletion of objs                      
 leo_comp_meta_with_dc_queue    |   idling    | 0              | 1600           | 500            | compare metadata w/remote-node              
 leo_delete_dir_queue_1         |   idling    | 0              | 1600           | 500            | deletion bucket #1                          
 leo_delete_dir_queue_2         |   idling    | 0              | 1600           | 500            | deletion bucket #2                          
 leo_delete_dir_queue_3         |   idling    | 0              | 1600           | 500            | deletion bucket #3                          
 leo_delete_dir_queue_4         |   idling    | 0              | 1600           | 500            | deletion bucket #4                          
 leo_delete_dir_queue_5         |   idling    | 0              | 1600           | 500            | deletion bucket #5                          
 leo_delete_dir_queue_6         |   idling    | 0              | 1600           | 500            | deletion bucket #6                          
 leo_delete_dir_queue_7         |   idling    | 0              | 1600           | 500            | deletion bucket #7                          
 leo_delete_dir_queue_8         |   idling    | 0              | 1600           | 500            | deletion bucket #8                          
 leo_per_object_queue           |   running   | 35             | 1600           | 500            | recover inconsistent objs                   
 leo_rebalance_queue            |   idling    | 0              | 1600           | 500            | rebalance objs                              
 leo_recovery_node_queue        |   idling    | 0              | 1600           | 500            | recovery objs of node                       
 leo_req_delete_dir_queue       |   running   | 4              | 1600           | 500            | request removing directories                
 leo_sync_by_vnode_id_queue     |   idling    | 0              | 1600           | 500            | sync objs by vnode-id                       
 leo_sync_obj_with_dc_queue     |   idling    | 0              | 1600           | 500            | sync objs w/remote-node                     
```
I think it should be possible to make node01 start by stopping cluster and deleting all delete-related queues on all nodes but I haven't tried that yet.

There are plans to upgrade to latest 1.4.1 version but I don't think it will be done until the situation is resolved, unless there are no other options.

**Comment by vstax** [comment]
Correction: node01 (the last launched node) has 6 AVS directories with 64 containers in each. All other nodes in cluster have 16 containers in each directory. There were plans to change them to 64 containers as well eventually after whole cluster is running on physical hardware.

So this operation (takeover from node03 to node01) essentially changed number of containers on new node as well, which might have something to do with the problem.

**Comment by mocchira** [comment]
@vstax Thanks for the additional info, we have started the investigation and planed to do the followings 

- Provides the procedure to fix the inconsistent RING problem
    - This procedure might include the operation on remote_console if it's impossible to fix the problem without remote_console.
    - Once the RING get fixed on all nodes, you'd have to issue recover-node on all nodes to recover the lack of replicas of each file stored in the cluster
- Fix the rebalance logic to make it atomic operation
    - From what @plmikhail provided here: https://github.com/leo-project/leofs/issues/1076#issuecomment-405076921, part of rebalance seems to fail and this might cause the root problem.
- Consider implementing sort of force update RING to deal with this kind of situation

**Comment by vstax** [comment]
@mocchira 
Thanks! Please also advise on how to make node01 start again (wiping all queues on the rest of the nodes or some other solution), because right now it still cannot be started.

**Comment by mocchira** [comment]
@vstax 

> Thanks! Please also advise on how to make node01 start again (wiping all queues on the rest of the nodes or some other solution), because right now it still cannot be started.

OK, Got it. let me clarify which task is the most high priority for you? maybe this one (node01 start again) ? I'm now assuming the each priority as below.

1. Get node01 back to running
2. Sync RING across the cluster somehow (then recover-node on all nodes)
3. Find out/Fix the root problem that caused the RING inconsistency on the take over operation
4. Implement the new command on leofs-adm to force sync RING across the cluster

Any suggestions/corrections are appreciated.

**Comment by vstax** [comment]
@mocchira 
I think that would be:
1. Sync RING
2. Get node01 back to running
3. Run recover-node on all nodes
4. Replace node09 (in VM) with the node running on physical hardware, using takeover procedure

Recover-node across cluster should be done after node01 is running, but making it running before RING is fixed is not very useful: the original problem of writes being refused will return. After node01 was launched (and RING broke) very high % of writes started to fail with an error, despite W=1 setting. Suspending or stopping node01 makes these writes work. The cluster is under production load so failing writes are a big problem.

Fixing root problem should probably be before or after 4), at least in a view of planned switch from 16 AVS files per directory to 64 (using takeover of all nodes), I think?

**Comment by mocchira** [comment]
@vstax OK. We'd like to get started as below

### Trouble Shooting
1. Sync RING
  - @mocchira (and/or other project members) will try to find out the way to do so
2. Get node01 back to running
  - @mocchira (ditto)
3. Run recover-node on all nodes
  - @vstax or @plmikhail 
4. Replace node09 (in VM) with the node running on physical hardware, using takeover procedure
  - @vstax or @plmikhail 
### Prevention of Recurrence
5. Find out/Fix the root problem that caused the RING inconsistency on the take over operation
  - @mocchira
6. Implement the new command on leofs-adm to force sync RING across the cluster
  - @mocchira

**Comment by mocchira** [comment]
> Correction: node01 (the last launched node) has 6 AVS directories with 64 containers in each. All other nodes in cluster have 16 containers in each directory. There were plans to change them to 64 containers as well eventually after whole cluster is running on physical hardware.

> So this operation (takeover from node03 to node01) essentially changed number of containers on new node as well, which might have something to do with the problem.

This can cause the errors (noproc) like followings

```erlang
[I]     node01@10.3.15.101      2018-07-14 17:44:01.73437 +0300 1531579441      leo_storage_handler_del_directory:terminate/2   232     [{"terminate with reason",{noproc,{gen_server,call,[undefined,{get,<<"develop/test/27.mp4\n4\n">>},30000]}}}]
[I]     node01@10.3.15.101      2018-07-14 17:44:01.106658 +0300        1531579441      leo_storage_handler_del_directory:terminate/2   232     [{"terminate with reason",{noproc,{gen_server,call,[undefined,{get,<<"develop/test/26.mp4\n3\n">>},30000]}}}]
```

due to the number of containers changed so that rolling obj.num_of_containers back to the original one might help node01 back to running.

**Comment by mocchira** [comment]
@vstax let me tell you the current status

> 1. Sync RING

@yosukehara is working on this and he will provide you with the procedure to get the inconsistent RING back to the normal status. Please wait for a while.

> 2. Get node01 back to running

I'm taking on this and have already answered at https://github.com/leo-project/leofs/issues/1076#issuecomment-406138897 . let me know whether or not it works

> Find out/Fix the root problem that caused the RING inconsistency on the take over operation

I'm taking on this and found out the reproducible step just now so I'll file the another issue for that. In short, Trying to join the node which node name was previously used by another node can cause this RING inconsistent problem. In your case, reusing node01@10.3.15.101 did kick the problem unfortunately. The reason why this causes the RING inconsistency  are still under investigation.

> Implement the new command on leofs-adm to force sync RING across the cluster

@yosukehara have some ideas to deal with the RING problem so he or I will file other issues to enable users to re-generate the proper RING across the cluster in an intuitive/easy way.

**Comment by mocchira** [comment]
filed https://github.com/leo-project/leofs/issues/1078 to solve the root problem

**Comment by vstax** [comment]
@mocchira 
Thanks for finding this out. I have a few questions:

> due to the number of containers changed so that rolling obj.num_of_containers back to the original one might help node01 back to running.

I don't quite understand what you meant here. The current node01 was launched with [64,64,64,64,64,64] num_of_containers from the very start. And rebalance operation filled all these 64 containers in each AVS directories with data. There doesn't seem to be any meaningful rollback, unless you mean kill all the data and recreate node01 with the same num_of_containers as all other nodes in the cluster (but what good would that do?).

Also, rebalance operation was going fine and finished (but the problem of lots of writes being refused persisted during rebalance). The node01 was stopped and restarted at least twice during rebalance to check if it would change anything. It was restarting just fine back then. Only during later experiments, after rebalance was finished, but the situation did not improve other things were tried, like "suspend" and "recover-ring" and one of them broke startup of node01.

> Trying to join the node which node name was previously used by another node can cause this RING inconsistent problem. In your case, reusing node01@10.3.15.101 did kick the problem unfortunately

Great find! I wonder if this would work: detach and kill current node01 (not sure if it would be better to keep the data in AVS or not?), rename it to some new name that was never used and attach it like that, then rebalance. Will this fix the RING? And if the RING will be fixed, will executing `recover-node` for whole cluster after that make sure that no data is lost?

Actually, there is one thing I don't know / understand yet, about how recover-node works. Imagine this scenario. We have some storage nodes (S0, S1, S2, S3, ...). By current RING, the object A goes to the node S0 (+Sx +Sy). We don't care about nodes Sx and Sy here, maybe the object was lost there, or whatever. Now, something happens to the RING (some rebalances or it broke or whatever, doesn't matter), now neither the current RING nor the old one make the object A go to S0, it should be on S1 + Sx' + Sy' instead. Let's assume no rebalance/recover is going on anymore at this point. This means that read of object A cannot complete as well, but it's no big deal.
Now we execute `recover-node S1`. Node S0 gets that request and starts to gather objects to be pushed to S1. The question is: will it push the object A, which (neither by current not by old) RING should be present on S0 to the correct node, S1, and thus will eventually recover it (after `recover-node Sx'` and `recover-node Sy'` is executed as well) or will it ignore it completely and A will be "lost" forever, despite being physically present on S0?

Because if the answer is true, current cluster in this installation can be just recreated with whole new identities in the worst case - which is a pretty bad option since lots of reads and writes will fail in process, but no current data will be lost - and execution of `recover-node` for each node will make all the data available again.

**Comment by mocchira** [comment]
@vstax 

> I don't quite understand what you meant here. The current node01 was launched with [64,64,64,64,64,64] num_of_containers from the very start. And rebalance operation filled all these 64 containers in each AVS directories with data. There doesn't seem to be any meaningful rollback, unless you mean kill all the data and recreate node01 with the same num_of_containers as all other nodes in the cluster (but what good would that do?).

Got it. I misunderstood the situation. so yes you are right as long as you are using 1.3.8. (with 1.4.0 or higher ones, users can issue recover-disk on leofs-adm ONLY in case every node has a same num_of_containers setting.)

> Also, rebalance operation was going fine and finished (but the problem of lots of writes being refused persisted during rebalance). The node01 was stopped and restarted at least twice during rebalance to check if it would change anything. It was restarting just fine back then. Only during later experiments, after rebalance was finished, but the situation did not improve other things were tried, like "suspend" and "recover-ring" and one of them broke startup of node01.

Thanks. I'd better understand the current situation.

Given that what you shared on the above comment and what I observed by reproducing this problem with https://github.com/leo-project/leofs/issues/1078, node01 would probably have the same RING with other storage nodes and have had files which are supposed to be stored in node01 so at this time I don't think any rollback is needed for node01.

so from now on

> Only during later experiments, after rebalance was finished, but the situation did not improve other things were tried, like "suspend" and "recover-ring" and one of them broke startup of node01.

I will try to come up with another solution to get node01 back to the cluster based on what you shared on the above comment.

> Great find! I wonder if this would work: detach and kill current node01 (not sure if it would be better to keep the data in AVS or not?), rename it to some new name that was never used and attach it like that, then rebalance. Will this fix the RING? And if the RING will be fixed, will executing recover-node for whole cluster after that make sure that no data is lost?

Maybe. I think it's worth trying so that I will try the above procedure on my dev-cluster later and report whether or not it works.

> Actually, there is one thing I don't know / understand yet, about how recover-node works. Imagine this scenario. We have some storage nodes (S0, S1, S2, S3, ...). By current RING, the object A goes to the node S0 (+Sx +Sy). We don't care about nodes Sx and Sy here, maybe the object was lost there, or whatever. Now, something happens to the RING (some rebalances or it broke or whatever, doesn't matter), now neither the current RING nor the old one make the object A go to S0, it should be on S1 + Sx' + Sy' instead. Let's assume no rebalance/recover is going on anymore at this point. This means that read of object A cannot complete as well, but it's no big deal.

> Now we execute recover-node S1. Node S0 gets that request and starts to gather objects to be pushed to S1. The question is: will it push the object A, which (neither by current not by old) RING should be present on S0 to the correct node, S1, and thus will eventually recover it (after recover-node Sx' and recover-node Sy' is executed as well) or will it ignore it completely and A will be "lost" forever, despite being physically present on S0?

I partly can't understand what you mean here so let me clarify a few things.

1. which nodes have **OLD** RING and which nodes have **NEW** RING? or all nodes have **NEW** RING? In mixed cases (some have OLD one and the other have NEW one), issuing the recover-node in the cluster with inconsistent RING might ignore(can't find the object due to the different RING) it so the recover-node will be in-complete however that doesn't mean A will be "lost" because there are some ways to recover RING to the **OLD**(proper) one which now @yosukehara is working on.

The only case which can cause the data loss is that 
- if there are no over wrapped nodes at any range between OLD and NEW RING
- issue recover-node during the above situation

(anyway in almost all cases, there should be at least one over wrapped node thanks to the consistent hashing nature unless nodes in the cluster have drastically been changed)

> Because if the answer is true, current cluster in this installation can be just recreated with whole new identities in the worst case - which is a pretty bad option since lots of reads and writes will fail in process, but no current data will be lost - and execution of recover-node for each node will make all the data available again.

Yes it's valid concern so we will compare the actual current RING status and the new RING status to be to check if there are some over wrapped nodes  at any range between two RINGs so this would avoid such worst case never happen.

**Comment by mocchira** [comment]
@vstax I summarized this problem at https://github.com/leo-project/leofs/issues/1078#issuecomment-406518483 so please check it out.

**Comment by mocchira** [comment]
@vstax Now it turns out that All Storage's RING table has been correct and all takeover operations has gone well so you can ignore any comments dropped at https://github.com/leo-project/leofs/issues/1076#issuecomment-406472111 to save your time, instead please check https://github.com/leo-project/leofs/issues/1078#issuecomment-406518483 out as described above.

**Comment by mocchira** [comment]
@vstax 

> I've tried clearing queues on node01 but it does not help it to start. There are "delete" messages in queues of other nodes, e.g.

Can you check that you had deleted the queues for delete-(bucket|directory) on node01?
The default path is ${QUEUE_DIR}/del_dir/state.
On my dev-box, it's placed as below
```bash
leofs@cat2neat:leofs.test$ ls -al package/leo_storage_3/work/queue/
total 52
drwxrwxr-x 13 leofs leofs 4096 Jul 19 14:49 .
drwxrwxr-x  5 leofs leofs 4096 Jul 19 15:06 ..
drwxrwxr-x 10 leofs leofs 4096 Jul 19 14:49 1
drwxrwxr-x 10 leofs leofs 4096 Jul 19 14:49 11
drwxrwxr-x 10 leofs leofs 4096 Jul 19 14:49 12
drwxrwxr-x 10 leofs leofs 4096 Jul 19 14:49 2
drwxrwxr-x 10 leofs leofs 4096 Jul 19 14:49 3
drwxrwxr-x 10 leofs leofs 4096 Jul 19 14:49 4
drwxrwxr-x 10 leofs leofs 4096 Jul 19 14:49 5
drwxrwxr-x 10 leofs leofs 4096 Jul 19 14:49 6
drwxrwxr-x 10 leofs leofs 4096 Jul 19 14:49 7
drwxrwxr-x  4 leofs leofs 4096 Jul 19 14:49 del_dir
drwxrwxr-x  5 leofs leofs 4096 Jul 19 14:49 membership
leofs@cat2neat:leofs.test$ ls -al package/leo_storage_3/work/queue/del_dir/
total 24
drwxrwxr-x  4 leofs leofs 4096 Jul 19 14:49 .
drwxrwxr-x 13 leofs leofs 4096 Jul 19 14:49 ..
lrwxrwxrwx  1 leofs leofs   87 Jul 19 14:49 state0 -> /home/leofs/dev/leofs.test/package/leo_storage_3/work/queue/del_dir/state0_63699198570/
drwxrwxr-x  9 leofs leofs 4096 Jul 20 08:46 state0_63699198570
lrwxrwxrwx  1 leofs leofs   87 Jul 19 14:49 state1 -> /home/leofs/dev/leofs.test/package/leo_storage_3/work/queue/del_dir/state1_63699198570/
drwxrwxr-x  9 leofs leofs 4096 Jul 20 08:46 state1_63699198570
```

From what I understand through reading the source code,

```
[I]     node01@10.3.15.101      2018-07-14 17:44:01.73437 +0300 1531579441      leo_storage_handler_del_directory:terminate/2   232     [{"terminate with reason",{noproc,{gen_server,call,[undefined,{get,<<"develop/test/27.mp4\n4\n">>},30000]}}}]
[I]     node01@10.3.15.101      2018-07-14 17:44:01.106658 +0300        1531579441      leo_storage_handler_del_directory:terminate/2   232     [{"terminate with reason",{noproc,{gen_server,call,[undefined,{get,<<"develop/test/26.mp4\n3\n">>},30000]}}}]
```

Those errors never happen in case ${QUEUE_DIR}/del_dir/state is properly created (automatically created during the boot process on a storage node)
So in almost all cases, this should solve the problem.
but I've just noticed there might be a racy problem that if

- leo_storage still doesn't initialize the leo_backend_db for delete-(bucket|directory)
- receive delete requests from other nodes

then the above errors can happen so as I commented on https://github.com/leo-project/leofs/issues/1078#issuecomment-406518483, deleting all queue files in the cluster and issuing mq-suspend leo_delete_dir_queue_[1-8] on all storage nodes except node01 then restarting node01 may work. Please give it a try.

**Comment by mocchira** [comment]
> but I've just noticed there might be a racy problem that if

This racy problem will be fixed by moving the leo_backend_db for delete-(bucket|dir) initializing logic (currently placed at https://github.com/leo-project/leofs/blob/v1/apps/leo_storage/src/leo_storage_app.erl#L236-L237) to the position before https://github.com/leo-project/leofs/blob/v1/apps/leo_storage/src/leo_storage_app.erl#L174 (this will start the rpc service so that any gen_server on a storage node start to receive requests)

I will file this issue later.

**Comment by vstax** [comment]
@mocchira 

> Can you check that you had deleted the queues for delete-(bucket|directory) on node01?
> The default path is ${QUEUE_DIR}/del_dir/state.

Yes, I was removing del_dir as well (everything except for "membership" subdirectory, in fact). Maybe it would be of help: del_dir was **NOT** recreated. The startup fails without creating this directory. Here is exactly what I did
```
cd /usr/local/leofs/current/leo_storage/work
cp -a queue queue.bak
cd queue
rm -rf ? ?? del_dir
systemctl start leofs-storage
```
At the end of startup node fails with messages like above, and here is output of `ls /usr/local/leofs/current/leo_storage/work/queue` after that:
`1  11  12  2  3  4  5  6  7  membership`

**Comment by vstax** [comment]
@mocchira 

> I partly can't understand what you mean here so let me clarify a few things.

Sorry that the question was not clear. It was probably too abstract from your point of view; it was just an example, what I really want to understand is how recover-node operation scans objects and whether it cares about anything except for new (current) RING.

> which nodes have OLD RING and which nodes have NEW RING? or all nodes have NEW RING?

Let's assume that yes, all nodes have new RING. Let me rephrase the question. From what I understand, what `recover-node Sx` essentially does is sending command to all other nodes (except for Sx) to go through AVS metadata and find all the objects that are supposed to be on Sx (according to current RING) and put them to the queue, processing of which will push these objects to Sx. My question was - is there some kind of filtering done during this scan through AVS metadata or not? The example for this question is: we have node S0, which is instructed to do recover-node for node Sx. It goes through objects that are on S0 and finds the object "A". According to the current RING, A definitely should be on Sx. But according both to current RING and to the old one (whether it's the same as current or differs from it), object "A" should not be on node S0. The question is: will this object be put to the queue for sending it to Sx or not? And if it will be put to the queue, will it actually be read from S0 even though no RING says that object "A" is on S0 right now - and actually pushed to S0 during queue processing?

**Comment by mocchira** [comment]
> Let's assume that yes, all nodes have new RING. Let me rephrase the question. From what I understand, what recover-node Sx essentially does is sending command to all other nodes (except for Sx) to go through AVS metadata and find all the objects that are supposed to be on Sx (according to current RING) and put them to the queue, processing of which will push these objects to Sx. My question was - is there some kind of filtering done during this scan through AVS metadata or not? The example for this question is: we have node S0, which is instructed to do recover-node for node Sx. It goes through objects that are on S0 and finds the object "A". According to the current RING, A definitely should be on Sx. But according both to current RING and to the old one (whether it's the same as current or differs from it), object "A" should not be on node S0. The question is: will this object be put to the queue for sending it to Sx or not? And if it will be put to the queue, will it actually be read from S0 even though no RING says that object "A" is on S0 right now - and actually pushed to S0 during queue processing?

OK. I think I got your point now.

Since LeoFS tries to find the requested file based on the current(for PUT/DELETE) or prev(for GET/HEAD) RING, so

> But according both to current RING and to the old one (whether it's the same as current or differs from it), object "A" should not be on node S0

With this assumption, 

> will this object be put to the queue for sending it to Sx or not?

It will NOT because LeoFS can't find the object "A" with (current/prev) RING.

> will it actually be read from S0 even though no RING says that object "A" is on S0 right now - and actually pushed to S0 during queue processing?

No it won't due to the same reason above.

**Comment by vstax** [comment]
@mocchira 
Thanks! I understand now (also why what you called "over wrapped nodes" are important here).

**Comment by mocchira** [comment]
@plmikhail @vstax I think the root problem is now filed on #1078 so we can close this issue so please do so if there are no remaining problems.
